cpp/src/arrow/flight/server.cc

```diff
@@ -47,145 +45,63 @@ using ServerWriter = grpc::ServerWriter<T>;
-// Implementation of ZeroCopyOutputStream that writes to a fixed-size buffer
-class FixedSizeProtoWriter : public ::google::protobuf::io::ZeroCopyOutputStream {
- public:
-  explicit FixedSizeProtoWriter(grpc_slice slice)
-      : slice_(slice),
-        bytes_written_(0),
-        total_size_(static_cast<int>(GRPC_SLICE_LENGTH(slice))) {}
-
-  bool Next(void** data, int* size) override {
-    // Consume the whole slice
-    *data = GRPC_SLICE_START_PTR(slice_) + bytes_written_;
-    *size = total_size_ - bytes_written_;
-    bytes_written_ = total_size_;
-    return true;
+#define CHECK_ARG_NOT_NULL(VAL, MESSAGE)                              \
+  if (VAL == nullptr) {                                               \
+    return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, MESSAGE); \
   }
 
-  void BackUp(int count) override { bytes_written_ -= count; }
-
-  int64_t ByteCount() const override { return bytes_written_; }
-
- private:
-  grpc_slice slice_;
-  int bytes_written_;
-  int total_size_;
-};
-
-// Write FlightData to a grpc::ByteBuffer without extra copying
-template <>
-class SerializationTraits<IpcPayload> {
+class ARROW_EXPORT FlightMessageReaderImpl : public FlightMessageReader {
  public:
-  static grpc::Status Deserialize(ByteBuffer* buffer, IpcPayload* out) {
-    return grpc::Status(grpc::StatusCode::UNIMPLEMENTED,
-                        "IpcPayload deserialization not implemented");
-  }
-
-  static grpc::Status Serialize(const IpcPayload& msg, ByteBuffer* out,
-                                bool* own_buffer) {
-    size_t total_size = 0;
-
-    DCHECK_LT(msg.metadata->size(), kInt32Max);
-    const int32_t metadata_size = static_cast<int32_t>(msg.metadata->size());
-
-    // 1 byte for metadata tag
-    total_size += 1 + WireFormatLite::LengthDelimitedSize(metadata_size);
-
-    int64_t body_size = 0;
-    for (const auto& buffer : msg.body_buffers) {
-      // Buffer may be null when the row length is zero, or when all
-      // entries are invalid.
-      if (!buffer) continue;
-
-      body_size += buffer->size();
-
-      const int64_t remainder = buffer->size() % 8;
-      if (remainder) {
-        body_size += 8 - remainder;
-      }
-    }
-
-    // 2 bytes for body tag
-    // Only written when there are body buffers
-    if (msg.body_length > 0) {
-      total_size +=
-          2 + WireFormatLite::LengthDelimitedSize(static_cast<size_t>(body_size));
-    }
-
-    // TODO(wesm): messages over 2GB unlikely to be yet supported
-    if (total_size > kInt32Max) {
-      return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
-                          "Cannot send record batches exceeding 2GB yet");
+  FlightMessageReaderImpl(const FlightDescriptor& descriptor,
+                          std::shared_ptr<Schema> schema,
+                          grpc::ServerReader<pb::FlightData>* reader)
+      : descriptor_{descriptor},
+        schema_{schema},
+        reader_{reader},
+        stream_finished_{false} {}
```

Comment:
Use parens in each of these initializations instead of braces to be more style-conforming